File: examples/Algorithm/python/03_stack_queue.py
"""
Stack and Queue Applications
Stack and Queue Applications

Examples of algorithms using stack (LIFO) and queue (FIFO) data structures.
"""

from collections import deque
from typing import List, Optional


# =============================================================================
# Stack Implementation (List-based)
# =============================================================================
class Stack:
    """List-based stack implementation"""

    def __init__(self):
        self._items = []

    def push(self, item):
        self._items.append(item)

    def pop(self):
        if not self.is_empty():
            return self._items.pop()
        raise IndexError("Stack is empty")

    def peek(self):
        if not self.is_empty():
            return self._items[-1]
        raise IndexError("Stack is empty")

    def is_empty(self):
        return len(self._items) == 0

    def size(self):
        return len(self._items)


# =============================================================================
# Queue Implementation (deque-based)
# =============================================================================
class Queue:
    """Deque-based queue implementation (O(1) operations)"""

    def __init__(self):
        self._items = deque()

    def enqueue(self, item):
        self._items.append(item)

    def dequeue(self):
        if not self.is_empty():
            return self._items.popleft()
        raise IndexError("Queue is empty")

    def front(self):
        if not self.is_empty():
            return self._items[0]
        raise IndexError("Queue is empty")

    def is_empty(self):
        return len(self._items) == 0

    def size(self):
        return len(self._items)


# =============================================================================
# 1. Valid Parentheses
# =============================================================================
def is_valid_parentheses(s: str) -> bool:
    """
    Validate parentheses string
    Time Complexity: O(n), Space Complexity: O(n)
    """
    stack = []
    mapping = {')': '(', '}': '{', ']': '['}

    for char in s:
        if char in mapping:
            # When encountering a closing bracket
            if not stack or stack[-1] != mapping[char]:
                return False
            stack.pop()
        else:
            # When encountering an opening bracket
            stack.append(char)

    return len(stack) == 0


# =============================================================================
# 2. Postfix Evaluation
# =============================================================================
def evaluate_postfix(expression: str) -> int:
    """
    Evaluate postfix expression
    Example: "2 3 + 4 *" = (2 + 3) * 4 = 20
    """
    stack = []
    operators = {'+', '-', '*', '/'}

    for token in expression.split():
        if token in operators:
            b = stack.pop()
            a = stack.pop()
            if token == '+':
                stack.append(a + b)
            elif token == '-':
                stack.append(a - b)
            elif token == '*':
                stack.append(a * b)
            elif token == '/':
                stack.append(int(a / b))  # Integer division
        else:
            stack.append(int(token))

    return stack[0]


# =============================================================================
# 3. Next Greater Element
# =============================================================================
def next_greater_element(arr: List[int]) -> List[int]:
    """
    Find the first greater element to the right of each element
    Returns -1 if none exists
    Time Complexity: O(n), Space Complexity: O(n)
    """
    n = len(arr)
    result = [-1] * n
    stack = []  # Store indices

    for i in range(n):
        # If current element is greater than stack top
        while stack and arr[i] > arr[stack[-1]]:
            idx = stack.pop()
            result[idx] = arr[i]
        stack.append(i)

    return result


# =============================================================================
# 4. Daily Temperatures
# =============================================================================
def daily_temperatures(temperatures: List[int]) -> List[int]:
    """
    Calculate how many days to wait until a warmer day
    Time Complexity: O(n), Space Complexity: O(n)
    """
    n = len(temperatures)
    result = [0] * n
    stack = []  # Store (index)

    for i in range(n):
        while stack and temperatures[i] > temperatures[stack[-1]]:
            prev_idx = stack.pop()
            result[prev_idx] = i - prev_idx
        stack.append(i)

    return result


# =============================================================================
# 5. Min Stack
# =============================================================================
class MinStack:
    """
    Stack that returns the minimum value in O(1) time
    """

    def __init__(self):
        self.stack = []
        self.min_stack = []

    def push(self, val: int):
        self.stack.append(val)
        if not self.min_stack or val <= self.min_stack[-1]:
            self.min_stack.append(val)

    def pop(self):
        if self.stack:
            val = self.stack.pop()
            if val == self.min_stack[-1]:
                self.min_stack.pop()

    def top(self) -> int:
        return self.stack[-1] if self.stack else None

    def get_min(self) -> int:
        return self.min_stack[-1] if self.min_stack else None


# =============================================================================
# 6. Stack Using Two Queues
# =============================================================================
class StackUsingQueues:
    """Implement a stack using two queues"""

    def __init__(self):
        self.q1 = deque()
        self.q2 = deque()

    def push(self, x: int):
        self.q2.append(x)
        while self.q1:
            self.q2.append(self.q1.popleft())
        self.q1, self.q2 = self.q2, self.q1

    def pop(self) -> int:
        return self.q1.popleft() if self.q1 else None

    def top(self) -> int:
        return self.q1[0] if self.q1 else None

    def empty(self) -> bool:
        return len(self.q1) == 0


# =============================================================================
# 7. Sliding Window Maximum (Monotonic Queue)
# =============================================================================
def max_sliding_window(nums: List[int], k: int) -> List[int]:
    """
    Find the maximum in a sliding window of size k
    Uses Monotonic Decreasing Deque
    Time Complexity: O(n), Space Complexity: O(k)
    """
    result = []
    dq = deque()  # Store indices, values are monotonically decreasing

    for i in range(len(nums)):
        # Remove indices outside the window range
        while dq and dq[0] < i - k + 1:
            dq.popleft()

        # Remove values smaller than current (from the back)
        while dq and nums[dq[-1]] < nums[i]:
            dq.pop()

        dq.append(i)

        # Add maximum once the window is complete
        if i >= k - 1:
            result.append(nums[dq[0]])

    return result


# =============================================================================
# Tests
# =============================================================================
def main():
    print("=" * 60)
    print("Stack and Queue Application Examples")
    print("=" * 60)

    # 1. Valid Parentheses
    print("\n[1] Valid Parentheses")
    test_cases = ["()", "()[]{}", "(]", "([)]", "{[]}"]
    for tc in test_cases:
        result = is_valid_parentheses(tc)
        print(f"    '{tc}' -> {result}")

    # 2. Postfix Evaluation
    print("\n[2] Postfix Evaluation")
    expressions = ["2 3 +", "2 3 + 4 *", "5 1 2 + 4 * + 3 -"]
    for expr in expressions:
        result = evaluate_postfix(expr)
        print(f"    '{expr}' = {result}")

    # 3. Next Greater Element
    print("\n[3] Next Greater Element")
    arr = [4, 5, 2, 25]
    result = next_greater_element(arr)
    print(f"    Array: {arr}")
    print(f"    Result: {result}")

    # 4. Daily Temperatures
    print("\n[4] Daily Temperatures")
    temps = [73, 74, 75, 71, 69, 72, 76, 73]
    result = daily_temperatures(temps)
    print(f"    Temperatures: {temps}")
    print(f"    Days to wait: {result}")

    # 5. Min Stack
    print("\n[5] Min Stack (MinStack)")
    min_stack = MinStack()
    operations = [
        ("push", 3), ("push", 5), ("getMin", None),
        ("push", 2), ("push", 1), ("getMin", None),
        ("pop", None), ("getMin", None)
    ]
    for op, val in operations:
        if op == "push":
            min_stack.push(val)
            print(f"    push({val})")
        elif op == "pop":
            min_stack.pop()
            print(f"    pop()")
        elif op == "getMin":
            print(f"    getMin() = {min_stack.get_min()}")

    # 6. Sliding Window Maximum
    print("\n[6] Sliding Window Maximum")
    nums = [1, 3, -1, -3, 5, 3, 6, 7]
    k = 3
    result = max_sliding_window(nums, k)
    print(f"    Array: {nums}, k={k}")
    print(f"    Window maximums: {result}")

    print("\n" + "=" * 60)


if __name__ == "__main__":
    main()

File: examples/Algorithm/cpp/03_stack_queue.cpp
/*
 * Stack and Queue
 * Stack, Queue, Deque, Monotonic Stack/Queue
 *
 * Applications of linear data structures.
 */

#include <iostream>
#include <vector>
#include <stack>
#include <queue>
#include <deque>
#include <string>
#include <unordered_map>

using namespace std;

// =============================================================================
// 1. Basic Stack Applications
// =============================================================================

// Valid parentheses check
bool isValidParentheses(const string& s) {
    stack<char> st;
    unordered_map<char, char> pairs = {{')', '('}, {']', '['}, {'}', '{'}};

    for (char c : s) {
        if (c == '(' || c == '[' || c == '{') {
            st.push(c);
        } else {
            if (st.empty() || st.top() != pairs[c])
                return false;
            st.pop();
        }
    }

    return st.empty();
}

// Reverse Polish Notation evaluation
int evalRPN(const vector<string>& tokens) {
    stack<int> st;

    for (const string& token : tokens) {
        if (token == "+" || token == "-" || token == "*" || token == "/") {
            int b = st.top(); st.pop();
            int a = st.top(); st.pop();

            if (token == "+") st.push(a + b);
            else if (token == "-") st.push(a - b);
            else if (token == "*") st.push(a * b);
            else st.push(a / b);
        } else {
            st.push(stoi(token));
        }
    }

    return st.top();
}

// Infix to Postfix conversion
string infixToPostfix(const string& infix) {
    stack<char> st;
    string postfix;
    unordered_map<char, int> precedence = {{'+', 1}, {'-', 1}, {'*', 2}, {'/', 2}};

    for (char c : infix) {
        if (isalnum(c)) {
            postfix += c;
        } else if (c == '(') {
            st.push(c);
        } else if (c == ')') {
            while (!st.empty() && st.top() != '(') {
                postfix += st.top();
                st.pop();
            }
            st.pop();  // Remove '('
        } else {
            while (!st.empty() && st.top() != '(' &&
                   precedence[st.top()] >= precedence[c]) {
                postfix += st.top();
                st.pop();
            }
            st.push(c);
        }
    }

    while (!st.empty()) {
        postfix += st.top();
        st.pop();
    }

    return postfix;
}

// =============================================================================
// 2. Monotonic Stack
// =============================================================================

// Next Greater Element
vector<int> nextGreaterElement(const vector<int>& nums) {
    int n = nums.size();
    vector<int> result(n, -1);
    stack<int> st;  // Store indices

    for (int i = 0; i < n; i++) {
        while (!st.empty() && nums[st.top()] < nums[i]) {
            result[st.top()] = nums[i];
            st.pop();
        }
        st.push(i);
    }

    return result;
}

// Largest Rectangle in Histogram
int largestRectangleArea(const vector<int>& heights) {
    stack<int> st;
    int maxArea = 0;
    int n = heights.size();

    for (int i = 0; i <= n; i++) {
        int h = (i == n) ? 0 : heights[i];

        while (!st.empty() && heights[st.top()] > h) {
            int height = heights[st.top()];
            st.pop();
            int width = st.empty() ? i : i - st.top() - 1;
            maxArea = max(maxArea, height * width);
        }
        st.push(i);
    }

    return maxArea;
}

// Daily Temperatures
vector<int> dailyTemperatures(const vector<int>& temperatures) {
    int n = temperatures.size();
    vector<int> result(n, 0);
    stack<int> st;

    for (int i = 0; i < n; i++) {
        while (!st.empty() && temperatures[st.top()] < temperatures[i]) {
            result[st.top()] = i - st.top();
            st.pop();
        }
        st.push(i);
    }

    return result;
}

// =============================================================================
// 3. Queue Applications
// =============================================================================

// Queue for BFS (simple example)
vector<int> bfsOrder(const vector<vector<int>>& graph, int start) {
    vector<int> result;
    vector<bool> visited(graph.size(), false);
    queue<int> q;

    q.push(start);
    visited[start] = true;

    while (!q.empty()) {
        int node = q.front();
        q.pop();
        result.push_back(node);

        for (int neighbor : graph[node]) {
            if (!visited[neighbor]) {
                visited[neighbor] = true;
                q.push(neighbor);
            }
        }
    }

    return result;
}

// =============================================================================
// 4. Deque Applications
// =============================================================================

// Sliding Window Maximum
vector<int> maxSlidingWindow(const vector<int>& nums, int k) {
    deque<int> dq;  // Store indices
    vector<int> result;

    for (int i = 0; i < (int)nums.size(); i++) {
        // Remove elements outside the window
        while (!dq.empty() && dq.front() <= i - k) {
            dq.pop_front();
        }

        // Remove elements smaller than current
        while (!dq.empty() && nums[dq.back()] < nums[i]) {
            dq.pop_back();
        }

        dq.push_back(i);

        if (i >= k - 1) {
            result.push_back(nums[dq.front()]);
        }
    }

    return result;
}

// =============================================================================
// 5. Queue Implementation Using Two Stacks
// =============================================================================

class MyQueue {
private:
    stack<int> input, output;

    void transfer() {
        if (output.empty()) {
            while (!input.empty()) {
                output.push(input.top());
                input.pop();
            }
        }
    }

public:
    void push(int x) {
        input.push(x);
    }

    int pop() {
        transfer();
        int val = output.top();
        output.pop();
        return val;
    }

    int peek() {
        transfer();
        return output.top();
    }

    bool empty() {
        return input.empty() && output.empty();
    }
};

// =============================================================================
// 6. Min Stack
// =============================================================================

class MinStack {
private:
    stack<int> st;
    stack<int> minSt;

public:
    void push(int val) {
        st.push(val);
        if (minSt.empty() || val <= minSt.top()) {
            minSt.push(val);
        }
    }

    void pop() {
        if (st.top() == minSt.top()) {
            minSt.pop();
        }
        st.pop();
    }

    int top() {
        return st.top();
    }

    int getMin() {
        return minSt.top();
    }
};

// =============================================================================
// Test
// =============================================================================

void printVector(const vector<int>& v) {
    cout << "[";
    for (size_t i = 0; i < v.size(); i++) {
        cout << v[i];
        if (i < v.size() - 1) cout << ", ";
    }
    cout << "]";
}

int main() {
    cout << "============================================================" << endl;
    cout << "Stack and Queue Examples" << endl;
    cout << "============================================================" << endl;

    // 1. Parentheses validation
    cout << "\n[1] Valid Parentheses" << endl;
    cout << "    \"()[]{}\" : " << (isValidParentheses("()[]{}") ? "valid" : "invalid") << endl;
    cout << "    \"([)]\"   : " << (isValidParentheses("([)]") ? "valid" : "invalid") << endl;

    // 2. Reverse Polish Notation
    cout << "\n[2] Reverse Polish Notation" << endl;
    vector<string> rpn = {"2", "1", "+", "3", "*"};
    cout << "    [\"2\",\"1\",\"+\",\"3\",\"*\"] = " << evalRPN(rpn) << endl;

    cout << "    \"a+b*c\" -> \"" << infixToPostfix("a+b*c") << "\"" << endl;

    // 3. Monotonic Stack
    cout << "\n[3] Monotonic Stack" << endl;
    vector<int> nums = {2, 1, 2, 4, 3};
    cout << "    Array: [2,1,2,4,3]" << endl;
    cout << "    Next greater element: ";
    printVector(nextGreaterElement(nums));
    cout << endl;

    vector<int> heights = {2, 1, 5, 6, 2, 3};
    cout << "    Histogram [2,1,5,6,2,3] largest rectangle: " << largestRectangleArea(heights) << endl;

    vector<int> temps = {73, 74, 75, 71, 69, 72, 76, 73};
    cout << "    Daily temperatures [73,74,75,71,69,72,76,73]: ";
    printVector(dailyTemperatures(temps));
    cout << endl;

    // 4. Sliding Window Maximum
    cout << "\n[4] Sliding Window Maximum" << endl;
    vector<int> window = {1, 3, -1, -3, 5, 3, 6, 7};
    cout << "    Array: [1,3,-1,-3,5,3,6,7], k=3" << endl;
    cout << "    Maximum values: ";
    printVector(maxSlidingWindow(window, 3));
    cout << endl;

    // 5. Queue using two stacks
    cout << "\n[5] Queue Using Two Stacks" << endl;
    MyQueue q;
    q.push(1);
    q.push(2);
    cout << "    push(1), push(2)" << endl;
    cout << "    peek(): " << q.peek() << endl;
    cout << "    pop(): " << q.pop() << endl;
    cout << "    empty(): " << (q.empty() ? "true" : "false") << endl;

    // 6. Min Stack
    cout << "\n[6] Min Stack" << endl;
    MinStack ms;
    ms.push(-2);
    ms.push(0);
    ms.push(-3);
    cout << "    push(-2), push(0), push(-3)" << endl;
    cout << "    getMin(): " << ms.getMin() << endl;
    ms.pop();
    cout << "    After pop() getMin(): " << ms.getMin() << endl;

    // 7. Complexity Summary
    cout << "\n[7] Complexity Summary" << endl;
    cout << "    | Operation    | Stack | Queue | Deque |" << endl;
    cout << "    |--------------|-------|-------|-------|" << endl;
    cout << "    | push/enqueue | O(1)  | O(1)  | O(1)  |" << endl;
    cout << "    | pop/dequeue  | O(1)  | O(1)  | O(1)  |" << endl;
    cout << "    | peek/front   | O(1)  | O(1)  | O(1)  |" << endl;

    cout << "\n============================================================" << endl;

    return 0;
}
